pkg/config/tlscfg/reload.go

```diff
@@ -0,0 +1,137 @@
+// Copyright (c) 2020 The Jaeger Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tlscfg
+
+import (
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"path/filepath"
+	"sync"
+
+	"github.com/fsnotify/fsnotify"
+	"go.uber.org/zap"
+)
+
+type watchCerts struct {
+	opts    Options
+	watcher *fsnotify.Watcher
+	cert    tls.Certificate
+	logger  *zap.Logger
+	mu      *sync.Mutex
+}
+
+var _ io.Closer = (*watchCerts)(nil)
+
+func newWatchCerts(opts Options, logger *zap.Logger) (*watchCerts, error) {
+	var cert tls.Certificate
+	if opts.CertPath != "" && opts.KeyPath != "" {
+		// load certs at startup to catch missing certs error early
+		c, err := tls.LoadX509KeyPair(filepath.Clean(opts.CertPath), filepath.Clean(opts.KeyPath))
+		if err != nil {
+			return nil, fmt.Errorf("failed to load server TLS cert and key: %w", err)
+		}
+		cert = c
+	}
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+	if err := addCertsToWatch(watcher, opts); err != nil {
+		return nil, err
+	}
+	return &watchCerts{
+		cert:    cert,
+		opts:    opts,
+		watcher: watcher,
+		logger:  logger,
+		mu:      &sync.Mutex{},
+	}, nil
+}
+
+func (w watchCerts) Close() error {
+	return w.watcher.Close()
+}
+
+func (w watchCerts) certificate() *tls.Certificate {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+	return &w.cert
+}
+
+func (w watchCerts) watchChangesLoop(rootCAs, clientCAs *x509.CertPool) {
+	for {
+		select {
+		case event, ok := <-w.watcher.Events:
+			if !ok {
+				return
+			}
+			w.logger.Info("Loading modified certificate",
+				zap.String("certificate", event.Name),
+				zap.String("event", event.Op.String()))
+			var err error
+			switch event.Name {
+			case w.opts.CAPath:
+				err = addCertToPool(w.opts.CAPath, rootCAs)
+			case w.opts.ClientCAPath:
+				err = addCertToPool(w.opts.ClientCAPath, clientCAs)
+			case w.opts.CertPath, w.opts.KeyPath:
+				w.mu.Lock()
+				w.cert, err = tls.LoadX509KeyPair(filepath.Clean(w.opts.CertPath), filepath.Clean(w.opts.KeyPath))
+				w.mu.Unlock()
+			}
+			if err != nil {
+				w.logger.Error("Failed to load certificate",
+					zap.String("certificate", event.Name),
+					zap.String("event", event.Op.String()),
+					zap.Error(err))
+			}
+		case err, ok := <-w.watcher.Errors:
+			if !ok {
+				return
+			}
+			w.logger.Error("Watcher got error", zap.Error(err))
```

Comment:
Also not sure what kind of errors it would get. If those are high-frequency as well, it would probably be better to have them at debug level.

Reply:
I would keep this at error level for now until we find out that these are high-freq (I haven't seen this being logged yet).